cwl/src/main/scala/cwl/WorkflowStepInputExpression.scala

```diff
@@ -23,18 +24,65 @@ final case class WorkflowStepInputExpression(input: WorkflowStepInput,
 
   override def evaluateValue(inputValues: Map[String, WomValue], ioFunctionSet: IoFunctionSet) = {
 
-    def lookupValue(key: String): Checked[WomValue] =
+    def lookupValue(key: String): ErrorOr[WomValue] =
       inputValues.
         get(FullyQualifiedName(key).id).
-        toRight(s"source value $key not found in input values ${inputValues.mkString("\n")}.  Graph Inputs were ${graphInputs.mkString("\n")}" |> NonEmptyList.one)
+        toValidNel(s"source value $key not found in input values ${inputValues.mkString("\n")}.  Graph Inputs were ${graphInputs.mkString("\n")}")
 
-    (input.valueFrom, input.source.map(_.fold(WorkflowStepInputSourceToStrings))) match {
-      case (None, Some(List(id))) =>
-        lookupValue(id).toValidated
+    (input.valueFrom, input.source.map(_.fold(StringOrStringArrayToStringList)), input.effectiveLinkMerge) match {
+
+      //When we have a single source, simply look it up
+      case (None, Some(List(source)), LinkMergeMethod.MergeNested) =>
+        lookupValue(source)
+
+      //When we have several sources, validate they are all present and provide them as a nested array of maps
+      case (None, Some(sources), LinkMergeMethod.MergeNested) =>
+
+        def convertTupleToWomMap: (String, WomValue) => WomMap = {
+          case (key, value) => WomMap(WomMapType(WomStringType, value.womType), Map(WomString(key) -> value))
+        }
```

Comment:
A function of no arguments that returns a two arg function that is only used after being `.tupled` into a function of one argument at the call site?